polaris/dataset/zarr/_manifest.py

```diff
@@ -25,6 +25,20 @@ def generate_zarr_manifest(zarr_root_path: str, output_dir: str):
     return zarr_manifest_path
 
 
+def deterministic_walk(root_path: str):
+    """
+    Recursively walk a directory and yield all files in a deterministic order.
+
+    Parameters:
+        root_path: The path to the root of the directory to walk
+    """
+
+    for dir_path, dirs, file_names in os.walk(root_path):
```

Comment:
The main blocker here is that `file_names` will be held in-memory for each directory that is traversed. Depending on the size of the dataset and the chunking mechanism/pattern chosen by the dataset creator, it is possible to have a directory which contains enough files at a single level to exhaust one's system memory. 

We actually [explored this idea](https://github.com/polaris-hub/polaris/pull/186/commits/a968e0be7081dd6134776dcf41c7a80282d2c2f4#r1742914766) during our first implementation. Apologies, I probably should've explained this more clearly in the issue.

Reply:
Ah no worries at all! I see why this would be problematic with all the `zarr` chunks. Feel free to close this PR as it won't address the issue. 

While researching for this issue I did come accross `zarr-checksum` that looks like it does what you need: https://github.com/dandi/zarr_checksum/tree/master and seems to have some of the `zarr` maintainers involved. If you think that looks like an acceptable solution then perhaps I could give that a try? Perhaps important question is whether the manifest file system is deployed (do we need to support two architectures?). Keen to hear your thoughts. 

I would love to keep contributing! If there are any issues I spy that I can help with. I also would love to contribute to XL competitions and benchmarks support (with my ASAPDiscovery hat on), so when the time comes I will try and help out.

Reply:
Building for the scale we hope to support has made for some interesting but tricky problems! It seems like Cas shared some additional context [here](https://github.com/polaris-hub/polaris/issues/188#issuecomment-2413631505) that I hope clarifies this checksum work a bit.

As for your other questions, the first version of the manifest file system is deployed and being used for V1 datasets which [leverage Zarr columns](https://polaris-hub.github.io/polaris/stable/tutorials/dataset_zarr.html). We're actively developing the second version of the system for V2 datasets, and it will diverge quite strongly in how it works relative to the current version. We plan to maintain both systems for the foreseeable future as deprecating the current system would make it challenging to ensure backward/forward compatibility.

Given this ongoing work, I'm going to close this PR, but we're looking forward to working together when you get the time! If you ever have any questions or need additional context, please feel free to reach out.